dev-tools/mage/gomod.go

```diff
@@ -0,0 +1,104 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mage
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/elastic/beats/dev-tools/mage/gotool"
+)
+
+// CopyModules contains a module name and the list of files or directories
+// to copy recursively.
+type CopyModules struct {
+	Name        string
+	FilesToCopy []string
+}
+
+var (
+	copyAll = []CopyModules{
+		CopyModules{
+			Name: "github.com/godror/godror",
+			FilesToCopy: []string{
+				"odpi",
+			},
+		},
+		CopyModules{
+			Name: "github.com/tsg/go-daemon",
+			FilesToCopy: []string{
+				"src",
+			},
+		},
+	}
+)
+
+// Vendor cleans up go.mod and copies the files not carried over from modules cache.
+func Vendor() error {
+	mod := gotool.Mod
+
+	err := mod.Tidy()
+	if err != nil {
+		return err
+	}
+
+	err = mod.Vendor()
+	if err != nil {
+		return err
+	}
+
+	err = mod.Verify()
+	if err != nil {
+		return err
+	}
+
+	repo, err := GetProjectRepoInfo()
+	if err != nil {
+		return err
+	}
+	vendorFolder := filepath.Join(repo.RootDir, "vendor")
+
+	// copy packages which require the whole tree
+	for _, p := range copyAll {
+		path, err := gotool.ListModulePath(p.Name)
+		if err != nil {
+			return err
+		}
+		if len(path) != 1 {
+			return fmt.Errorf("unexpected number of paths")
+		}
+
+		for _, f := range p.FilesToCopy {
+			from := filepath.Join(path[0], f)
+			to := filepath.Join(vendorFolder, p.Name, f)
+			copyTask := &CopyTask{Source: from, Dest: to, DirMode: os.ModeDir | 0750}
+			err = copyTask.Execute()
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// AddModuleFilesToCopy adds modules which contain files
+// which has to be copied to the folder vendor (e.g. C source).
+func AddModuleFilesToCopy(files []CopyModules) {
+	copyAll = append(copyAll, files...)
+}
```

Comment:
Do we need this function? I don't see it being called from anywhere at the moment. Perhaps we can add it later when we actually need it? Speculative Generality, YAGNI, etc.